File: apps/ios/GalaxySSI/AgentEvalBenchmarkCatalog.swift
import Foundation

enum AgentEvalBenchmarkCatalog {
  static let standard = AgentBenchmarkSuite(
    id: "galaxyssi-ios-real-agent",
    version: "1.2.0",
    title: "GalaxySSI iOS Real Agent EvalOps",
    cases: taskQualityCases + planningCases + iosWorldCases + immediateMemoryCases + recoveryCases + multiAgentCases,
    targetPassRate: 0.95
  )

  static let longitudinalMemory = AgentBenchmarkSuite(
    id: "galaxyssi-ios-longitudinal-memory",
    version: "1.0.0",
    title: "GalaxySSI iOS 30/90-day Memory Certification",
    cases: longTermMemoryCases,
    targetPassRate: 0.95,
    minimumTaskCount: 10,
    maximumTaskCount: 10
  )

  static let suites = [standard, longitudinalMemory]

  static func suite(id: String, version: String) -> AgentBenchmarkSuite? {
    suites.first { $0.id == id && $0.version == version }
  }

  private static let taskQualityCases: [AgentBenchmarkCase] = [
    quality("quality-01", "Multi-step arithmetic", "Calculate (17 x 23) - (144 / 12). Return only the final integer.", "^379$"),
    quality("quality-02", "Constrained sorting", "Sort 9, 2, 5, 2, 1 ascending. Return a JSON array without explanation.", "^\\[\\s*1\\s*,\\s*2\\s*,\\s*2\\s*,\\s*5\\s*,\\s*9\\s*\\]$"),
    qualityJSON("quality-03", "Structured output", "Return only a valid JSON object where status is ready and count is 3.", ["status": "ready", "count": "3"]),
    quality("quality-04", "Dependency path", "A takes 2 minutes. B follows A and takes 3. C follows A and takes 4. D waits for B and C and takes 1. Give the minimum duration and critical path.", "(?is)(7\\s*(minutes?|min)).*(A.*C.*D)"),
    quality("quality-05", "Weighted score", "Weights are 0.5, 0.3, 0.2 and scores are 80, 90, 70. Return only the weighted total.", "^81$"),
    quality("quality-06", "Conditional reasoning", "Every blue box is heavy. Box K is not heavy. Can K be blue? Answer no and give one sentence of reasoning.", "(?is)^no.*(blue.*heavy|not heavy.*blue)"),
    qualityJSON("quality-07", "Information extraction", "Extract device, os, and ram from: iPhone 15 Pro, iOS 18, 8GB. Return one JSON line.", ["device": "iPhone 15 Pro", "os": "iOS 18", "ram": "8GB"]),
    quality("quality-08", "Unit conversion", "How many MiB are in 1.5 GiB? Return only the number.", "^1536$"),
    quality("quality-09", "Distinct events", "For [a,b,a,c,b,d], return the distinct count and events in first-seen order.", "(?is)(4).*(a.*b.*c.*d)"),
    quality("quality-10", "Conflict resolution", "Record 1 says screen understanding is enabled. A newer record 2 says it was removed. What is the current state?", "(?is)(removed|disabled).*(record 2|newer|latest)")
  ]

  private static let planningCases: [AgentBenchmarkCase] = [
    planning("plan-tool-01", "Device version", "Plan briefly, use an available tool to verify this device model and iOS version, then report tool evidence."),
    planning("plan-tool-02", "App version", "Plan, use a tool to read the current GalaxySSI version and build, then report evidence."),
    planning("plan-tool-03", "Network state", "Plan, use a tool to check network availability and validation, then report evidence."),
    planning("plan-tool-04", "Battery state", "Plan, use a tool to read battery level, charging state, and thermal state, then report evidence."),
    planning("plan-tool-05", "Storage state", "Plan, use a tool to read available app storage, then report the value and evidence."),
    planning("plan-tool-06", "Trusted time", "Plan, call a trusted time tool for local date and time zone, then identify the tool source."),
    planning("plan-tool-07", "File integrity", "Plan, create a text file in an allowed test directory, read it, calculate SHA-256, then report the digest."),
    planning("plan-tool-08", "Tool recovery", "Plan, attempt a side-effect-free tool check, adjust the approach once if it fails, and report every step."),
    planning("plan-tool-09", "Primary research", "Plan, find two primary sources with a search tool, cross-check them, and return a sourced conclusion.", sources: true),
    planning("plan-tool-10", "Artifact validation", "Plan, generate a small JSON artifact, validate that a tool can parse it, and report the validation result.")
  ]

  private static let iosWorldCases: [AgentBenchmarkCase] = [
    world("ios-world-01", "Foreground app", "Verify that GalaxySSI is foreground and report the observed screen."),
    world("ios-world-02", "Visible title", "Read and report the current GalaxySSI screen title."),
    world("ios-world-03", "Bundle identity", "Verify and report the GalaxySSI application bundle identifier."),
    world("ios-world-04", "App version", "Read and report the installed GalaxySSI version."),
    world("ios-world-05", "App build", "Read and report the installed GalaxySSI build number."),
    world("ios-world-06", "Interface locale", "Read and report the current interface locale identifier."),
    world("ios-world-07", "Time zone", "Read and report the current time zone identifier."),
    world("ios-world-08", "Low power mode", "Read and report whether iOS Low Power Mode is enabled."),
    world("ios-world-09", "Device model", "Read and report this iOS device model."),
    world("ios-world-10", "System version", "Read and report the current iOS system version.")
  ]

  private static let immediateMemoryCases: [AgentBenchmarkCase] = [
    immediateMemory("immediate-memory-01", "Immediate identity", "Return the cross-session memory value for IM-01 only.", "GSSI-IM-NOVA"),
    immediateMemory("immediate-memory-02", "Immediate preference", "Return the cross-session memory value for IM-02 only.", "GSSI-IM-DARK"),
    immediateMemory("immediate-memory-03", "Immediate device", "Return the cross-session memory value for IM-03 only.", "GSSI-IM-PHONE"),
    immediateMemory("immediate-memory-04", "Immediate project", "Return the cross-session memory value for IM-04 only.", "GSSI-IM-PROJECT"),
    immediateMemory("immediate-memory-05", "Immediate knowledge", "Return the cross-session memory value for IM-05 only.", "GSSI-IM-KNOWLEDGE"),
    immediateMemory("immediate-memory-06", "Immediate workflow", "Return the cross-session memory value for IM-06 only.", "GSSI-IM-WORKFLOW"),
    immediateMemory("immediate-memory-07", "Immediate decision", "Return the cross-session memory value for IM-07 only.", "GSSI-IM-DECISION"),
    immediateMemory("immediate-memory-08", "Memory update", "Return the current IM-08 value, not the superseded value.", "GSSI-IM-CURRENT", "GSSI-IM-OLD"),
    immediateMemory("immediate-memory-09", "Entity disambiguation", "Return only the value belonging to IM-09-B, not IM-09-A.", "GSSI-IM-BETA", "GSSI-IM-ALPHA"),
    immediateMemory("immediate-memory-10", "Source tracing", "Return the provenance-linked cross-session memory value for IM-10 only.", "GSSI-IM-PROVENANCE")
  ]

  private static let longTermMemoryCases: [AgentBenchmarkCase] = [
    memory("memory-30-01", "30-day identity", "Return the value of memory fixture M30-01 without guessing.", 30, "GSSI-M30-ALPHA"),
    memory("memory-30-02", "30-day preference", "Return the value of memory fixture M30-02 without guessing.", 30, "GSSI-M30-BRAVO"),
    memory("memory-30-03", "30-day device", "Return the value of memory fixture M30-03 without guessing.", 30, "GSSI-M30-CHARLIE"),
    memory("memory-30-04", "30-day project", "Return the value of memory fixture M30-04 without guessing.", 30, "GSSI-M30-DELTA"),
    memory("memory-30-05", "30-day decision", "Return the value of memory fixture M30-05 without guessing.", 30, "GSSI-M30-ECHO"),
    memory("memory-90-01", "90-day identity", "Return the value of memory fixture M90-01 without guessing.", 90, "GSSI-M90-FOXTROT"),
    memory("memory-90-02", "90-day preference", "Return the value of memory fixture M90-02 without guessing.", 90, "GSSI-M90-GOLF"),
    memory("memory-90-03", "90-day device", "Return the value of memory fixture M90-03 without guessing.", 90, "GSSI-M90-HOTEL"),
    memory("memory-90-04", "90-day project", "Return the value of memory fixture M90-04 without guessing.", 90, "GSSI-M90-INDIA"),
    memory("memory-90-05", "90-day update", "Return the current value of memory fixture M90-05, not its retired value.", 90, "GSSI-M90-JULIET")
  ]

  private static let recoveryCases: [AgentBenchmarkCase] = [
    recovery("recovery-network-01", "Network recovery 1", "Complete the reliability echo after one network interruption. Return RECOVERED-NET-01 exactly once.", .networkLoss, "RECOVERED-NET-01"),
    recovery("recovery-network-02", "Network recovery 2", "Checkpoint the task and recover after a network interruption. Return RECOVERED-NET-02 exactly once.", .networkLoss, "RECOVERED-NET-02"),
    recovery("recovery-network-03", "Network recovery 3", "Execute a recoverable task across network loss. Return RECOVERED-NET-03 exactly once.", .networkLoss, "RECOVERED-NET-03"),
    recovery("recovery-process-01", "Process recovery 1", "Checkpoint before simulated process termination. Return RECOVERED-PROC-01 exactly once after recovery.", .processDeath, "RECOVERED-PROC-01"),
    recovery("recovery-process-02", "Process recovery 2", "Recover this task after simulated process termination. Return RECOVERED-PROC-02 exactly once.", .processDeath, "RECOVERED-PROC-02"),
    recovery("recovery-process-03", "Process recovery 3", "Remain idempotent across simulated process termination. Return RECOVERED-PROC-03 exactly once.", .processDeath, "RECOVERED-PROC-03"),
    recovery("recovery-doze-01", "Background recovery 1", "Recover after iOS background suspension. Return RECOVERED-IDLE-01 exactly once.", .doze, "RECOVERED-IDLE-01"),
    recovery("recovery-doze-02", "Background recovery 2", "Checkpoint and recover after iOS background suspension. Return RECOVERED-IDLE-02 exactly once.", .doze, "RECOVERED-IDLE-02"),
    recovery("recovery-reboot-01", "Restart recovery 1", "Recover after a device restart. Return RECOVERED-BOOT-01 exactly once.", .reboot, "RECOVERED-BOOT-01"),
    recovery("recovery-reboot-02", "Restart recovery 2", "Remain idempotent across a device restart. Return RECOVERED-BOOT-02 exactly once.", .reboot, "RECOVERED-BOOT-02")
  ]

  private static let multiAgentCases: [AgentBenchmarkCase] = [
    multi("multi-agent-01", "Implement and review", "Use two distinct Agents: one proposes an implementation and one independently reviews it, then merge the conclusion."),
    multi("multi-agent-02", "Research and challenge", "Use two distinct Agents: one researches and one seeks counterexamples, then merge the conclusion."),
    multi("multi-agent-03", "Plan and risk", "Use two distinct Agents: one decomposes a plan and one reviews risks, then merge the conclusion."),
    multi("multi-agent-04", "Code and test", "Use two distinct Agents: one designs code and one independently designs tests, then merge the conclusion."),
    multi("multi-agent-05", "Performance and UX", "Use two distinct Agents: one analyzes performance and one analyzes UX, then prioritize findings."),
    multi("multi-agent-06", "Security dual review", "Use two distinct Agents to independently assess one security design and identify agreement and disagreement."),
    multi("multi-agent-07", "Three-Agent review", "Use three distinct Agents for proposal, verification, and challenge, then return one synthesis.", agents: 3),
    multi("multi-agent-08", "Blind model review", "Use two distinct Agents independently, compare their hidden-identity results, and explain the selection."),
    multi("multi-agent-09", "Failure diagnosis", "Use two distinct Agents: one diagnoses root cause and one verifies the fix against regression."),
    multi("multi-agent-10", "Evidence merge", "Use two distinct Agents to provide evidence, deduplicate it, and return one traceable conclusion.")
  ]

  private static func quality(_ id: String, _ title: String, _ prompt: String, _ pattern: String) -> AgentBenchmarkCase {
    AgentBenchmarkCase(id: id, dimension: .taskQuality, title: title, prompt: prompt,
      expectation: AgentBenchmarkExpectation(requiredOutputPatterns: [pattern]))
  }

  private static func qualityJSON(
    _ id: String,
    _ title: String,
    _ prompt: String,
    _ fields: [String: String]
  ) -> AgentBenchmarkCase {
    AgentBenchmarkCase(id: id, dimension: .taskQuality, title: title, prompt: prompt,
      expectation: AgentBenchmarkExpectation(requiredJsonFields: fields))
  }

  private static func planning(_ id: String, _ title: String, _ prompt: String, sources: Bool = false) -> AgentBenchmarkCase {
    var evidence: Set<AgentOutcomeEvidenceKind> = [.finalResponse, .toolReceipt]
    if sources { evidence.insert(.verifiedSource) }
    return AgentBenchmarkCase(id: id, dimension: .planningAndTools, title: title, prompt: prompt,
      expectation: AgentBenchmarkExpectation(minimumOutputCharacters: 20, minimumPlanEvents: 1,
        minimumToolReceipts: 1, minimumVerifiedSources: sources ? 2 : 0, requiredEvidence: evidence))
  }

  private static func immediateMemory(
    _ id: String,
    _ title: String,
    _ prompt: String,
    _ expected: String,
    _ forbidden: String = ""
  ) -> AgentBenchmarkCase {
    AgentBenchmarkCase(id: id, dimension: .immediateMemory, title: title, prompt: prompt,
      expectation: AgentBenchmarkExpectation(
        requiredOutputPatterns: [NSRegularExpression.escapedPattern(for: expected)],
        forbiddenOutputPatterns: forbidden.isEmpty ? [] : [NSRegularExpression.escapedPattern(for: forbidden)],
        requiredEvidence: [.finalResponse, .memoryProvenance]
      ))
  }

  private static func world(_ id: String, _ title: String, _ prompt: String) -> AgentBenchmarkCase {
    AgentBenchmarkCase(id: id, dimension: .iosWorld, title: title, prompt: prompt,
      expectation: AgentBenchmarkExpectation(requiredEvidence: [.finalResponse, .programmaticVerifier],
        iosWorldTaskId: id, requireIOSObservedValuesInOutput: true))
  }

  private static func memory(_ id: String, _ title: String, _ prompt: String, _ days: Int, _ expected: String) -> AgentBenchmarkCase {
    AgentBenchmarkCase(id: id, dimension: .longTermMemory, title: title,
      prompt: "\(prompt) Use only provenance-linked memory written at least \(days) days ago.",
      expectation: AgentBenchmarkExpectation(requiredOutputPatterns: [NSRegularExpression.escapedPattern(for: expected)],
        requiredEvidence: [.finalResponse, .memoryProvenance], memoryHorizonDays: days))
  }

  private static func recovery(_ id: String, _ title: String, _ prompt: String, _ condition: AgentEvalCondition, _ expected: String) -> AgentBenchmarkCase {
    let escaped = NSRegularExpression.escapedPattern(for: expected)
    return AgentBenchmarkCase(id: id, dimension: .recovery, title: title, prompt: prompt,
      expectation: AgentBenchmarkExpectation(requiredOutputPatterns: [escaped],
        forbiddenOutputPatterns: ["(?s)(\(escaped).*){2,}"], requiredEvidence: [.finalResponse, .recoveryEvent],
        requiredCondition: condition))
  }

  private static func multi(_ id: String, _ title: String, _ prompt: String, agents: Int = 2) -> AgentBenchmarkCase {
    AgentBenchmarkCase(id: id, dimension: .multiAgent, title: title, prompt: prompt,
      expectation: AgentBenchmarkExpectation(minimumOutputCharacters: 40, minimumDistinctAgents: agents,
        minimumHandoffs: agents - 1))
  }
}

File: apps/android/app/src/main/java/com/signalasi/chat/AgentEvalBenchmarkCatalog.kt
package com.signalasi.chat

object AgentEvalBenchmarkCatalog {
    val standard: AgentBenchmarkSuite by lazy {
        AgentBenchmarkSuite(
            id = "signalasi-android-real-agent",
            version = "1.2.0",
            title = "SignalASI Android Real Agent EvalOps",
            cases = taskQualityCases() + planningCases() + androidWorldCases() +
                immediateMemoryCases() + recoveryCases() + multiAgentCases(),
            targetPassRate = 0.95
        )
    }

    val longitudinalMemory: AgentBenchmarkSuite by lazy {
        AgentBenchmarkSuite(
            id = "signalasi-android-longitudinal-memory",
            version = "1.0.0",
            title = "SignalASI Android 30/90-day Memory Certification",
            cases = longTermMemoryCases(),
            targetPassRate = 0.95,
            minimumTaskCount = 10,
            maximumTaskCount = 10
        )
    }

    val suites: List<AgentBenchmarkSuite> by lazy { listOf(standard, longitudinalMemory) }

    fun suite(id: String, version: String): AgentBenchmarkSuite? = suites.firstOrNull {
        it.id == id && it.version == version
    }

    private fun taskQualityCases() = listOf(
        quality("quality-01", "多步算术", "计算 (17 × 23) - (144 ÷ 12)，只给出最终整数。", "^379$"),
        quality("quality-02", "约束排序", "将 9、2、5、2、1 升序排列，使用 JSON 数组输出，不要解释。", "^\\[\\s*1\\s*,\\s*2\\s*,\\s*2\\s*,\\s*5\\s*,\\s*9\\s*]$"),
        qualityJson(
            "quality-03",
            "结构化输出",
            "只输出有效 JSON 对象，status 必须为 ready，count 必须为 3。",
            mapOf("status" to "ready", "count" to "3")
        ),
        quality("quality-04", "依赖路径", "任务 A 用2分钟；B在A后用3分钟；C在A后用4分钟；D要等B和C后用1分钟。给出最短总时长和关键路径。", "(?s)(7\\s*分钟|7\\s*min).*(A.*C.*D)"),
        quality("quality-05", "加权评分", "三个指标权重为0.5、0.3、0.2，得分为80、90、70。计算加权总分，只输出数字。", "^81$"),
        quality("quality-06", "条件推理", "所有蓝盒都很重；盒子K不重。K是否可能是蓝盒？只回答‘不可能’并给出一句理由。", "(?s)^不可能.*(蓝盒.*重|不重.*蓝盒)"),
        qualityJson(
            "quality-07",
            "信息抽取",
            "从‘设备SM-T575，Android 13，内存4GB’提取 device、os、ram，只输出一行JSON。",
            mapOf("device" to "SM-T575", "os" to "Android 13", "ram" to "4GB")
        ),
        quality("quality-08", "单位换算", "1.5 GiB 等于多少 MiB？只输出数字。", "^1536$"),
        quality("quality-09", "去重统计", "事件序列[a,b,a,c,b,d]中有多少个不同事件？列出数量和按首次出现顺序的事件。", "(?s)(4).*(a.*b.*c.*d)"),
        quality(
            "quality-10",
            "冲突识别",
            "记录1：屏幕理解已启用。记录2（时间更晚）：屏幕理解已移除。当前状态是什么？",
            "(?s)(?=.*(?:已移除|移除))(?=.*(?:记录2|更新|较晚|更晚|最新)).*"
        )
    )

    private fun planningCases() = listOf(
        planning("plan-tool-01", "设备版本核验", "先列出简短计划，再使用可用工具核验当前设备型号和 Android 版本，最后给出工具证据。"),
        planning("plan-tool-02", "应用版本核验", "先规划，再使用工具读取当前 SignalASI 的版本名称和版本号，最后报告证据。"),
        planning("plan-tool-03", "网络状态核验", "先规划，再使用工具检查当前网络是否可用及是否已验证，最后给出结论和证据。"),
        planning("plan-tool-04", "电池状态核验", "先规划，再使用工具读取电量、充电状态和温度，最后报告工具证据。"),
        planning("plan-tool-05", "存储状态核验", "先规划，再使用工具读取应用可用存储空间，最后给出数值和证据。"),
        planning("plan-tool-06", "时间核验", "先规划，再调用可信时间工具获取当前本地日期和时区，最后说明工具来源。"),
        planning("plan-tool-07", "文件完整性", "先规划，再在允许的测试目录创建文本文件、读取验证并计算 SHA-256，最后给出摘要。"),
        planning("plan-tool-08", "失败恢复", "先规划，尝试一个无副作用的工具检查；若失败则调整一次方案，最后报告每一步结果。"),
        planning("plan-tool-09", "多来源调研", "先规划，再使用检索工具查找两份一手资料，交叉验证后给出带来源的结论。", sources = true),
        planning("plan-tool-10", "产物验证", "先规划，生成一个小型 JSON 产物，使用工具验证其可解析性，最后报告验证结果。")
    )

    private fun androidWorldCases() = listOf(
        world("android-world-01", "前台应用", "确认 SignalASI 当前位于前台，并报告其包名。"),
        world("android-world-02", "Android ID", "读取并报告当前设备的 Android ID。"),
        world("android-world-03", "屏幕亮度", "读取并报告当前系统屏幕亮度数值。"),
        world("android-world-04", "息屏时间", "读取并报告当前系统息屏超时毫秒数。"),
        world("android-world-05", "飞行模式", "读取并报告当前飞行模式状态值。"),
        world("android-world-06", "无障碍状态", "读取并报告当前无障碍总开关状态值。"),
        world("android-world-07", "自动旋转", "读取并报告当前自动旋转状态值。"),
        world("android-world-08", "字体比例", "读取并报告当前系统字体比例。"),
        world("android-world-09", "定位模式", "读取并报告当前系统定位模式值。"),
        world("android-world-10", "省电模式", "读取并报告当前系统低电量模式状态值。")
    )

    private fun immediateMemoryCases() = listOf(
        immediateMemory("immediate-memory-01", "即时身份记忆", "从跨会话即时记忆回答 IM-01 的值；只输出记住的值。", "SASI-IM-NOVA"),
        immediateMemory("immediate-memory-02", "即时偏好记忆", "从跨会话即时记忆回答 IM-02 的值；只输出记住的值。", "SASI-IM-DARK"),
        immediateMemory("immediate-memory-03", "即时设备记忆", "从跨会话即时记忆回答 IM-03 的值；只输出记住的值。", "SASI-IM-TABLET"),
        immediateMemory("immediate-memory-04", "即时项目记忆", "从跨会话即时记忆回答 IM-04 的值；只输出记住的值。", "SASI-IM-PROJECT"),
        immediateMemory("immediate-memory-05", "即时知识记忆", "从跨会话即时记忆回答 IM-05 的值；只输出记住的值。", "SASI-IM-KNOWLEDGE"),
        immediateMemory("immediate-memory-06", "即时工作流记忆", "从跨会话即时记忆回答 IM-06 的值；只输出记住的值。", "SASI-IM-WORKFLOW"),
        immediateMemory("immediate-memory-07", "即时决策记忆", "从跨会话即时记忆回答 IM-07 的值；只输出记住的值。", "SASI-IM-DECISION"),
        immediateMemory("immediate-memory-08", "记忆更新", "回答 IM-08 当前有效的值；不要返回已经被替代的旧值。", "SASI-IM-CURRENT", "SASI-IM-OLD"),
        immediateMemory("immediate-memory-09", "实体消歧", "只回答属于 IM-09-B 的记忆值，不要混入 IM-09-A。", "SASI-IM-BETA", "SASI-IM-ALPHA"),
        immediateMemory("immediate-memory-10", "来源追踪", "从带来源的跨会话即时记忆回答 IM-10 的值；只输出记住的值。", "SASI-IM-PROVENANCE")
    )

    private fun longTermMemoryCases() = listOf(
        memory("memory-30-01", "30天身份记忆", "从长期记忆回答测试夹具 M30-01 的值；不要猜测。", 30, "SASI-M30-ALPHA"),
        memory("memory-30-02", "30天偏好记忆", "从长期记忆回答测试夹具 M30-02 的值；不要猜测。", 30, "SASI-M30-BRAVO"),
        memory("memory-30-03", "30天设备记忆", "从长期记忆回答测试夹具 M30-03 的值；不要猜测。", 30, "SASI-M30-CHARLIE"),
        memory("memory-30-04", "30天项目记忆", "从长期记忆回答测试夹具 M30-04 的值；不要猜测。", 30, "SASI-M30-DELTA"),
        memory("memory-30-05", "30天决策记忆", "从长期记忆回答测试夹具 M30-05 的值；不要猜测。", 30, "SASI-M30-ECHO"),
        memory("memory-90-01", "90天身份记忆", "从长期记忆回答测试夹具 M90-01 的值；不要猜测。", 90, "SASI-M90-FOXTROT"),
        memory("memory-90-02", "90天偏好记忆", "从长期记忆回答测试夹具 M90-02 的值；不要猜测。", 90, "SASI-M90-GOLF"),
        memory("memory-90-03", "90天设备记忆", "从长期记忆回答测试夹具 M90-03 的值；不要猜测。", 90, "SASI-M90-HOTEL"),
        memory("memory-90-04", "90天项目记忆", "从长期记忆回答测试夹具 M90-04 的值；不要猜测。", 90, "SASI-M90-INDIA"),
        memory("memory-90-05", "90天更新记忆", "从长期记忆回答测试夹具 M90-05 当前有效的值；不要返回已废弃值。", 90, "SASI-M90-JULIET")
    )

    private fun immediateMemory(
        id: String,
        title: String,
        prompt: String,
        expected: String,
        forbidden: String = ""
    ) = AgentBenchmarkCase(
        id,
        AgentBenchmarkDimension.IMMEDIATE_MEMORY,
        title,
        prompt,
        AgentBenchmarkExpectation(
            requiredOutputPatterns = listOf(Regex.escape(expected)),
            forbiddenOutputPatterns = forbidden.takeIf(String::isNotBlank)?.let(Regex::escape)?.let(::listOf)
                ?: emptyList(),
            requiredEvidence = setOf(
                AgentOutcomeEvidenceKind.FINAL_RESPONSE,
                AgentOutcomeEvidenceKind.MEMORY_PROVENANCE
            )
        )
    )

    private fun recoveryCases() = listOf(
        recovery("recovery-network-01", "断网恢复1", "完成可靠性回声任务；测试期间会发生一次断网，恢复后只返回一次 RECOVERED-NET-01。", AgentEvalCondition.NETWORK_LOSS, "RECOVERED-NET-01"),
        recovery("recovery-network-02", "断网恢复2", "保存检查点并完成任务；测试期间会发生网络中断，恢复后只返回一次 RECOVERED-NET-02。", AgentEvalCondition.NETWORK_LOSS, "RECOVERED-NET-02"),
        recovery("recovery-network-03", "断网恢复3", "执行可恢复任务；测试期间会断网，恢复后只返回一次 RECOVERED-NET-03。", AgentEvalCondition.NETWORK_LOSS, "RECOVERED-NET-03"),
        recovery("recovery-process-01", "进程死亡恢复1", "保存检查点；测试期间会发生进程死亡，恢复后只返回一次 RECOVERED-PROC-01。", AgentEvalCondition.PROCESS_DEATH, "RECOVERED-PROC-01"),
        recovery("recovery-process-02", "进程死亡恢复2", "执行可恢复任务；测试期间会杀进程，恢复后只返回一次 RECOVERED-PROC-02。", AgentEvalCondition.PROCESS_DEATH, "RECOVERED-PROC-02"),
        recovery("recovery-process-03", "进程死亡恢复3", "保持幂等；测试期间会进程死亡，恢复后只返回一次 RECOVERED-PROC-03。", AgentEvalCondition.PROCESS_DEATH, "RECOVERED-PROC-03"),
        recovery("recovery-doze-01", "Doze恢复1", "测试期间设备会进入 Doze，退出后只返回一次 RECOVERED-DOZE-01。", AgentEvalCondition.DOZE, "RECOVERED-DOZE-01"),
        recovery("recovery-doze-02", "Doze恢复2", "保存检查点并等待设备休眠，退出待机后只返回一次 RECOVERED-DOZE-02。", AgentEvalCondition.DOZE, "RECOVERED-DOZE-02"),
        recovery("recovery-reboot-01", "重启恢复1", "测试期间设备会重启，恢复后只返回一次 RECOVERED-BOOT-01。", AgentEvalCondition.REBOOT, "RECOVERED-BOOT-01"),
        recovery("recovery-reboot-02", "重启恢复2", "保持任务幂等；测试期间会重启手机，恢复后只返回一次 RECOVERED-BOOT-02。", AgentEvalCondition.REBOOT, "RECOVERED-BOOT-02")
    )

    private fun multiAgentCases() = listOf(
        multi("multi-agent-01", "实现与审查", "至少让两个不同 Agent 协作：一个提出实现方案，一个独立审查，最后合并结论。"),
        multi("multi-agent-02", "研究与反证", "至少让两个不同 Agent 协作：一个给出研究结论，一个寻找反例，最后合并结论。"),
        multi("multi-agent-03", "计划与风险", "至少让两个不同 Agent 协作：一个拆解计划，一个审查风险，最后合并结论。"),
        multi("multi-agent-04", "编码与测试", "至少让两个不同 Agent 协作：一个设计代码，一个独立设计测试，最后合并结论。"),
        multi("multi-agent-05", "性能与UX", "至少让两个不同 Agent 协作：一个分析性能，一个分析UX，最后合并优先级。"),
        multi("multi-agent-06", "安全双审", "至少让两个不同 Agent 独立分析同一安全方案，指出共识与分歧。"),
        multi("multi-agent-07", "双阶段评审", "让两个不同 Agent 分别负责方案与独立反驳，最后只给一份经过复核的综合结论。"),
        multi("multi-agent-08", "模型盲测", "至少让两个不同 Agent 独立回答，再隐藏身份比较结果并给出选择依据。"),
        multi("multi-agent-09", "故障诊断", "至少让两个不同 Agent 协作：一个定位根因，一个验证修复不会回归。"),
        multi("multi-agent-10", "证据合并", "至少让两个不同 Agent 各自提供证据，去重并输出可追溯的统一结论。")
    )

    private fun quality(id: String, title: String, prompt: String, pattern: String) = AgentBenchmarkCase(
        id, AgentBenchmarkDimension.TASK_QUALITY, title, prompt,
        AgentBenchmarkExpectation(requiredOutputPatterns = listOf(pattern), minimumOutputChars = 1)
    )

    private fun qualityJson(
        id: String,
        title: String,
        prompt: String,
        fields: Map<String, String>
    ) = AgentBenchmarkCase(
        id, AgentBenchmarkDimension.TASK_QUALITY, title, prompt,
        AgentBenchmarkExpectation(requiredJsonFields = fields, minimumOutputChars = 1)
    )

    private fun planning(id: String, title: String, prompt: String, sources: Boolean = false) = AgentBenchmarkCase(
        id, AgentBenchmarkDimension.PLANNING_AND_TOOLS, title, prompt,
        AgentBenchmarkExpectation(
            minimumOutputChars = 20,
            minimumPlanEvents = 1,
            minimumToolReceipts = 1,
            requiredEvidence = buildSet {
                add(AgentOutcomeEvidenceKind.FINAL_RESPONSE)
                add(AgentOutcomeEvidenceKind.TOOL_RECEIPT)
                if (sources) add(AgentOutcomeEvidenceKind.VERIFIED_SOURCE)
            },
            minimumVerifiedSources = if (sources) 2 else 0
        )
    )

    private fun world(id: String, title: String, prompt: String) = AgentBenchmarkCase(
        id, AgentBenchmarkDimension.ANDROID_WORLD, title, prompt,
        AgentBenchmarkExpectation(
            minimumOutputChars = 1,
            requiredEvidence = setOf(
                AgentOutcomeEvidenceKind.FINAL_RESPONSE,
                AgentOutcomeEvidenceKind.PROGRAMMATIC_VERIFIER
            ),
            androidWorldTaskId = id,
            requireAndroidObservedValuesInOutput = true
        )
    )

    private fun memory(id: String, title: String, prompt: String, days: Int, expected: String) = AgentBenchmarkCase(
        id,
        AgentBenchmarkDimension.LONG_TERM_MEMORY,
        title,
        "$prompt 只允许使用至少${days}天前写入且带来源的记忆。",
        AgentBenchmarkExpectation(
            requiredOutputPatterns = listOf(Regex.escape(expected)),
            memoryHorizonDays = days,
            requiredEvidence = setOf(
                AgentOutcomeEvidenceKind.FINAL_RESPONSE,
                AgentOutcomeEvidenceKind.MEMORY_PROVENANCE
            )
        )
    )

    private fun recovery(
        id: String,
        title: String,
        prompt: String,
        condition: AgentEvalCondition,
        expected: String
    ) = AgentBenchmarkCase(
        id, AgentBenchmarkDimension.RECOVERY, title, prompt,
        AgentBenchmarkExpectation(
            requiredOutputPatterns = listOf(Regex.escape(expected)),
            forbiddenOutputPatterns = listOf("(?s)(${Regex.escape(expected)}.*){2,}"),
            requiredCondition = condition,
            requiredEvidence = setOf(
                AgentOutcomeEvidenceKind.FINAL_RESPONSE,
                AgentOutcomeEvidenceKind.RECOVERY_EVENT
            )
        )
    )

    private fun multi(id: String, title: String, prompt: String, agents: Int = 2) = AgentBenchmarkCase(
        id, AgentBenchmarkDimension.MULTI_AGENT, title, prompt,
        AgentBenchmarkExpectation(
            minimumOutputChars = 40,
            minimumDistinctAgents = agents,
            minimumHandoffs = agents - 1,
            requiredEvidence = setOf(AgentOutcomeEvidenceKind.FINAL_RESPONSE)
        )
    )
}
